chainerx_cc/chainerx/device.cc

```diff
@@ -68,9 +77,16 @@ ApplyBatchNormResult ApplyBatchNorm(
         CHAINERX_ASSERT(mean.GetTotalSize() == reduced_total_size);
         CHAINERX_ASSERT(var.GetTotalSize() == reduced_total_size);
     }
-    Array inv_std = Reciprocal(Sqrt(var + eps));
+    // TODO(hvy): Avoid AsType.
+    const Array& x_cast = x.AsType(dtype, false);
+    const Array& gamma_cast = gamma.AsType(dtype, false);
+    const Array& beta_cast = beta.AsType(dtype, false);
+    const Array& mean_cast = mean.AsType(dtype, false);
+    const Array& var_cast = var.AsType(dtype, false);
+
+    Array inv_std = Reciprocal(Sqrt(var_cast + eps));
 
-    Array out = (x - mean) * inv_std * gamma + beta;
+    Array out = (x_cast - mean_cast) * inv_std * gamma_cast + beta_cast;
```

Comment:
Why not cast `out` to the output dtype here?
It's confusing if `ApplyBatchNormResult::out.dtype()` is different from the final dtype.